<!DOCTYPE html>
<html>
  <head>
    <meta http-equiv="Content-Type" content="text/html;charset=utf-8">
    <title>WASP Result View</title>
    <script type="text/javascript" src="../../d3.v2.js"></script>
    <link type="text/css" href="tree-interactive.css" rel="stylesheet" />
  </head>
  
  <body>
  <div id="container">
	<div id="left-container">
		<div id="treeview"></div>
	</div>

	<div id="right-container">
		<div id="Job-div">
			<h4>Job Description</h4>
			<ul>
				<li><b>Job Name</b>
					<ul>
						<li>chip_sb </li>
					</ul>
				</li>
				<li><b>Assay Type</b>
					<ul>
						<li>ChIP-SEQ </li>
					</ul>
				</li>
				<li><b>Submitted By</b>
					<ul>
						<li>Amit Verma (Chanin 302B Verma)</li>
					</ul>
				</li>
				<li><b>Submitted Date</b>
					<ul>
						<li>04/23/12 </li>
					</ul>
				</li>
				<li><b>Data Last Updated</b>
					<ul>
						<li>06/14/12 </li>
					</ul>
				</li>
				<li><b>Software Versions</b>
					<ul>
						<li>WASP Pipeline
							<ul>
								<li>WASP - 3.0.55 (rev. 6278)</li>
							</ul>
						</li>
						<li>Peak Finder 
							<ul>
								<li>MACS - 1.4.2 </li>
							</ul>
						</li>
						<li>Aligner
							<ul>
								<li>bowtie - 0.12.7 </li>
							</ul>
						</li>
					</ul>
				</li>
				<li><b>Job Status</b>
					<ul>
						<li><span style="color: green">COMPLETE</span>
						</li>
					</ul>
				</li>
				<li><a href="http://wasp.einstein.yu.edu/results/production_wiki/AVerma/AVerma/P406/J10469/StdOutAndErr.txt">
					Processing and Analysis Software Logs</a>
				</li>
				<li><a href="http://wasp.einstein.yu.edu/results/production_wiki/AVerma/AVerma/P406/J10469/stats/stats.html" target="_blank">
					Sequencing Quality Metrics</a>
				</li>
			</ul>
		</div>
		
		<div id="Sample-div" style="display:none;">
			<h4>Sample Details</h4>
			<ul>
				<li><b>Sample Name</b>
					<ul>
						<li> AsPC1 Inp</li>
					</ul>
				</li>
				<li><b>Sample Type</b>
					<ul>
						<li>Input</li>
					</ul>
				</li>
				<li><b>Material</b>
					<ul>
						<li>Chromatin</li>
					</ul>
				</li>
				<li><b>Single/Pair-end</b>
					<ul>
						<li>Single</li>
					</ul>
				</li>
				<li><b>Raw Data Files</b>
					<ul>
						<li><a href="http://wasp.einstein.yu.edu/results/production_wiki/AVerma/AVerma/P406/J10469/raw/TrueSeqUnknown.AD0W92ACXX.lane_1_P0_I0.hg19.sequence.fastq.bz2">Raw Reads (fastq.bz2 file)</a></li>
						<li><a href="http://wasp.einstein.yu.edu/stream/streamFastqPf.php?file=production_wiki/AVerma/AVerma/P406/J10469/raw/TrueSeqUnknown.AD0W92ACXX.lane_1_P0_I0.hg19.sequence.fastq.bz2">Passed Filter Reads (fastq.bz2 file)</a></li>
					</ul>
				</li>
				<li><a href="http://wasp.einstein.yu.edu/results/production_wiki/AVerma/AVerma/P406/J10469/stats/stats_AsPC1_Inp.AD0W92ACXX.lane_1_P0_I1.fastq.html" target="_blank">
					Sequencing Quality Metrics</a>
				</li>
			</ul>
		</div>
		
		<div id="Library-div" style="display:none;">
			<h4>Library Details</h4>
			<ul>
				<li><b>Library Name</b>
					<ul>
						<li> AsPC1 Inp</li>
					</ul>
				</li>
				<li><b>Library Size</b>
					<ul>
						<li>502bp</li>
					</ul>
				</li>
				<li><b>Library Created With</b>
					<ul>
						<li>TruSEQ INDEXED PE DNA (ILLUMINA)</li>
					</ul>
				</li>
				<li><b>Index Tag</b>
					<ul>
						<li>Index 2 (CGATGT)</li>
					</ul>
				</li>
				<li><b>3' Adaptor</b>
					<ul>
						<li>AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC</li>
					</ul>
				</li>
				<li><b>Generated By</b>
					<ul>
						<li>Kevin Lau</li>
					</ul>
				</li>
				<li><b>Library Final Status</b>
					<ul>
						<li><span style="color: blue">COMPLETE</span></li>
					</ul>
				</li>
			</ul>
		</div>
		
		
		<div id="Result-div" style="display:none;">
			<h4>Peak Finding Details</h4>
			<p>
				Peak finding is performed using MACS (Model-based Analysis for ChIP-Seq). <a href="http://liulab.dfci.harvard.edu/MACS/">Click here</a> to view the website, or refer to the publication below: <br>
			</p>
			<ul>
				<li>Zhang et al. Model-based Analysis of ChIP-Seq (MACS). Genome Biol (2008) vol. 9 (9) pp. R137).</li>
			</ul>
			<p>
				The input for MACS is the BAM (.bam) and BAM index (.bam.bai) files provided above. Default parameters are used for the most part. If you are re-running an analysis using MACS at the command line yourself, the automated analysis can be replicated as follows (explanations of all MACS parameters can be found <a href="http://liulab.dfci.harvard.edu/MACS/00README.html">here</a> and the ACTUAL values of the parameters used in this analysis can be found at the top of the '.xls information on called peaks' file linked to below): <br>
			</p>
			<ul>
				<li><b>IP only</b>
					<ul>
						<li>macs --treatment=&lt;bam file for IP sample&gt; --name=&lt;sample name&gt; --format=BAM --gsize=&lt;effective genome size&gt; --tsize=&lt;size of sequence fragments&gt; --bw=&lt;sonication fragment size * 0.5&gt; --wig --space=100</li>
					</ul>
				</li>
				<li><b>IP vs input</b>
					<ul>
						<li>macs --treatment=&lt;bam file for IP sample&gt; --control=&lt;bam file for INPUT sample&gt; --name=&lt;sample name&gt; --format=BAM --gsize=&lt;effective genome size&gt; --tsize=&lt;size of sequence fragments&gt; --bw=&lt;sonication fragment size * 0.5&gt; --wig --space=100</li>
					</ul>
				</li>
			</ul>
			<p><br></br></p>
			<ul>
				<li><b>Peak Finding Results:</b>
					<ul>
						<li><a href="http://genome.ucsc.edu/cgi-bin/hgTracks?db=hg19&amp;hgt.customText=http://wasp.einstein.yu.edu/results/production_wiki/AVerma/AVerma/P406/J10469/analyzed/2-5_PD.AD0W92ACXX.lane_2_P0_I21.hg19__2-5_Inp.AD0W92ACXX.lane_1_P0_I9.hg19/2-5_PD.AD0W92ACXX.lane_2_P0_I21.hg19__2-5_Inp.AD0W92ACXX.lane_1_P0_I9.hg19_peaks.bed.bz2" target="_blank">Show peaks in Genome Browser</a></li>
						<li><a href="http://genome.ucsc.edu/cgi-bin/hgTracks?db=hg19&amp;hgt.customText=http://wasp.einstein.yu.edu/results/production_wiki/AVerma/AVerma/P406/J10469/analyzed/2-5_PD.AD0W92ACXX.lane_2_P0_I21.hg19__2-5_Inp.AD0W92ACXX.lane_1_P0_I9.hg19/2-5_PD.AD0W92ACXX.lane_2_P0_I21.hg19__2-5_Inp.AD0W92ACXX.lane_1_P0_I9.hg19_summits.bed.bz2" target="_blank">Show summits in Genome Browser</a></li>
						<li><a href="http://genome.ucsc.edu/cgi-bin/hgTracks?db=hg19&amp;hgt.customText=http://wasp.einstein.yu.edu/results/production_wiki/AVerma/AVerma/P406/J10469/analyzed/2-5_PD.AD0W92ACXX.lane_2_P0_I21.hg19__2-5_Inp.AD0W92ACXX.lane_1_P0_I9.hg19/2-5_PD.AD0W92ACXX.lane_2_P0_I21.hg19__2-5_Inp.AD0W92ACXX.lane_1_P0_I9.hg19_IP_track.bw" target="_blank">Show wiggle track for IP in Genome Browser</a></li>
						<li><a href="http://genome.ucsc.edu/cgi-bin/hgTracks?db=hg19&amp;hgt.customText=http://wasp.einstein.yu.edu/results/production_wiki/AVerma/AVerma/P406/J10469/analyzed/2-5_PD.AD0W92ACXX.lane_2_P0_I21.hg19__2-5_Inp.AD0W92ACXX.lane_1_P0_I9.hg19/2-5_PD.AD0W92ACXX.lane_2_P0_I21.hg19__2-5_Inp.AD0W92ACXX.lane_1_P0_I9.hg19_INPUT_track.bw" target="_blank">Show wiggle track for INPUT in Genome Browser</a></li>
						<li><a href="http://wasp.einstein.yu.edu/results/production_wiki/AVerma/AVerma/P406/J10469/analyzed/2-5_PD.AD0W92ACXX.lane_2_P0_I21.hg19__2-5_Inp.AD0W92ACXX.lane_1_P0_I9.hg19/2-5_PD.AD0W92ACXX.lane_2_P0_I21.hg19__2-5_Inp.AD0W92ACXX.lane_1_P0_I9.hg19_peaks.xls">Click to download .xls information on called peaks</a></li>
						<li><a href="http://wasp.einstein.yu.edu/results/production_wiki/AVerma/AVerma/P406/J10469/analyzed/2-5_PD.AD0W92ACXX.lane_2_P0_I21.hg19__2-5_Inp.AD0W92ACXX.lane_1_P0_I9.hg19/2-5_PD.AD0W92ACXX.lane_2_P0_I21.hg19__2-5_Inp.AD0W92ACXX.lane_1_P0_I9.hg19_negative_peaks.xls">Click to download .xls information on negative peaks</a></li>
						<li><a href="http://wasp.einstein.yu.edu/results/production_wiki/AVerma/AVerma/P406/J10469/analyzed/2-5_PD.AD0W92ACXX.lane_2_P0_I21.hg19__2-5_Inp.AD0W92ACXX.lane_1_P0_I9.hg19/2-5_PD.AD0W92ACXX.lane_2_P0_I21.hg19__2-5_Inp.AD0W92ACXX.lane_1_P0_I9.hg19_model.pdf">Click to download PDF image of peak model</a></li>
					</ul>
				</li>
			</ul>
		</div>
	
	</div>

  </div>
  <script type="text/javascript" src="tree-interactive-changed.js"></script>
  </body>
</html>
